apps/src/applab/EventSandboxer.js

```diff
@@ -99,63 +99,68 @@ EventSandboxer.prototype.sandboxEvent = function(event) {
     }
   });
 
-  // map touch events to mouse events
-  if (event.type && event.type.substring(0, 5) === 'touch') {
-    switch (event.type) {
-      case 'touchmove':
-        newEvent.type = 'mousemove';
-        break;
-      case 'touchstart':
-        newEvent.type = 'mousedown';
-        break;
-      case 'touchend':
-        newEvent.type = 'mouseup';
-        break;
-    }
-    event.clientX = event.changedTouches[0].clientX;
-    event.clientY = event.changedTouches[0].clientY;
-    event.pageX = event.changedTouches[0].pageX;
-    event.pageY = event.changedTouches[0].pageY;
-    event.x = event.changedTouches[0].clientX;
-    event.y = event.changedTouches[0].clientY;
+  // This "mouse" event will come either from a normal input, or from the first
+  // of multiple touches.
+  let mouseEvent;
+
+  const touchEvents = {
+    touchstart: 'mousedown',
+    touchmove: 'mousemove',
+    touchend: 'mouseup'
+  };
+
+  if (touchEvents[event.type]) {
+    newEvent.type = touchEvents[event.type];
+    mouseEvent = event.changedTouches[0];
+  } else {
+    mouseEvent = event;
   }
 
   // Convert x coordinates and then pass through to applabEvent:
   ['clientX', 'pageX', 'x'].forEach(function(prop) {
-    if (typeof event[prop] !== 'undefined') {
-      newEvent[prop] = (event[prop] - this.xOffset_) / this.xScale_;
+    if (typeof mouseEvent[prop] !== 'undefined') {
+      newEvent[prop] = (mouseEvent[prop] - this.xOffset_) / this.xScale_;
     }
   }, this);
   // Convert y coordinates and then pass through to applabEvent:
   ['clientY', 'pageY', 'y'].forEach(function(prop) {
-    if (typeof event[prop] !== 'undefined') {
-      newEvent[prop] = (event[prop] - this.yOffset_) / this.yScale_;
+    if (typeof mouseEvent[prop] !== 'undefined') {
+      newEvent[prop] = (mouseEvent[prop] - this.yOffset_) / this.yScale_;
     }
   }, this);
+
+  // Add values that can be missing when touch is used.
+  if (mouseEvent.x === undefined) {
+    newEvent.x = (mouseEvent.pageX - this.xOffset_) / this.xScale_;
```

Comment:
Are we guaranteed to have a `pageX` in all cases? i.e. keyUp/keyDown? Suggestion: when you create `mouseEvent` from `event.changedTouches[0]` above, assign the x and y. i.e.  `mouseEvent.x = event.changedTouches[0].pageX`

Reply:
That's a good point, and I will rearrange the code . That said, once assigned a value, `mouseEvent` will be a [`TouchEvent`](https://developer.mozilla.org/en-US/docs/Web/API/Touch) or some other kind of [`Event`](https://developer.mozilla.org/en-US/docs/Web/API/Event) and I want to treat it as an immutable, read-only object as well.  I want to write only into `newEvent` which is the local object that is designed to receive all the calculated values.

In fact, it appears that MouseEvent.x is just an alias for MouseEvent.clientX, so we could modify the other App Lab code that depends on this value, but I want to keep this change localised.